Bug 1214392 denormalize stats

mathjazz opened the pull request:
@Osmose Since you worked on #217, I wanted to check with you if this isn't fundamentally flawed. No need for an in-depth review yet.

What does it do?
* Adds denormalized stats fields to Project, Locale and ProjectLocale objects (in addition to Stats).
* Populates these fields with data migration and on every translation save/delete/upload and sync.
* Refactors charts in templates a little bit (reuses code via macro).
* Removes unused code: Transifex fields and update_entity_count().

What remains to be done (some things could be deferred to a later PR)?
- This might be a good opportunity to rename Stats to TranslatedResource.
- Stats object should use same stats field names as other objects.
- I guess pontoon.sync.core:update_resources() will be slow.
- Chart calculations could be moved from template to view.
- We could update latest translation and stats in the same place.
- We should update latest translation and stats also on project delete/disable/changelocales (pontoon.base.signals).

Osmose commented on pontoon/base/models.py:
```diff
@@ -354,6 +358,12 @@ class Project(models.Model):
     has_changed = models.BooleanField(default=False)
 
+    # Aggregated stats for this project.
+    total_strings = models.PositiveIntegerField(default=0)
```
I've never been able to decide if aggregated fields should be prefixed with `aggregate_` or something. It's super-verbose, but it's also explicit.

Osmose commented on pontoon/base/models.py:
```diff
@@ -496,6 +515,42 @@ def get_latest_activity(cls, self, extra=None):
+        return chart
+
+    @classmethod
+    def get_chart_dict(cls, obj):
```
Does this need to be a dict? If not, maybe this method is unnecessary?

mathjazz replied:
The data is being JSON serialized in `*_selector.html` templates (needed for the tooltip), which is why I used the `dict`.

Osmose commented on pontoon/base/models.py:
```diff
@@ -1308,9 +1269,42 @@ def update_stats(resource, locale):
+    translated_strings_diff = translated - stats.translated_count
+
+    # Stats
+    stats.approved_count += approved_strings_diff
```
You'll want to use `F` objects to make this an atomic update.

Osmose commented on pontoon/base/views.py:
```diff
@@ -143,18 +140,38 @@ def update_group(group, name):
+
+    locales = (
+        project.locales.all()
+        .prefetch_related(
```
Smells like a queryset method!

Osmose commented:
A few minor comments, but the overall shape of this looks good. I think you're on the right track!

mathjazz commented:
I've addressed the comments, updated tests, and made some additional cleanups.

Stats are now also updated if locale is removed from project, project or locale is deleted, or project disabled status changes.

@Osmose r?

Osmose commented on pontoon/base/models.py:
```diff
@@ -451,6 +527,12 @@ class ProjectLocale(models.Model):
     locale = models.ForeignKey(Locale, related_name='project_locale')
 
+    # Aggregated stats for this project in this locale.
+    total_strings = models.PositiveIntegerField(default=0)
```
These fields and the operations around them smell like an abstract model base class you can reuse for all the models in question.

Osmose commented on pontoon/base/models.py:
```diff
@@ -1073,255 +1215,73 @@ class Stats(models.Model):
         on_delete=models.SET_NULL
     )
 
+    objects = TranslatedResourceQuerySet.as_manager()
```
`as_manager` is the bestest thing ever

Osmose commented:
I only had time to skim this for now; if you want a more detailed review, I can give one, it just will take a while to happen. But it looks good! Abstracting the denormalization out to a base class is the only major change I can see that would be worthwhile. Nice work!
